Feature/interactive physical layout

jbrekelmans opened the pull request:
1/3 Pull requests on repositories Netflix/vizceral, Netflix/vizceral-example and Netflix/vizceral-react with the following features around "optimizing layout manually":
1. Repositioning of nodes by dragging them
2. Interactive Particle System simulation (nodes are particles and edges are spring forces) because it's awesome and allow for good automatic layout on some graphs
3. Resetting layout if #2 get's out of hand
4. Layout algorithm that respects positional metadata in the JSON graph files and otherwise lays out nodes in a circle
5. Export (well, Console.log) current node positions to be used as positional metadata once a satisfactory manual layout has been created

How to run:
1. Clone the branch feature/interactive_physical_layout of repositories jbrekelmans/vizceral, jbrekelmans/vizceral-example and jbrekelmans/vizceral-react locally
2. Run command 'npm link' in vizceral
3. Run command 'npm link vizceral' in vizceral-react
4. Run command 'npm link' in vizceral-react
5. Run command 'npm link vizceral-react' in vizceral-example
6. Make sure a graph (e.g. in sample_data.json) with 10-20 nodes is loaded for best results using the default Particle System settings

aaronblohowiak commented:
Very interesting! 🤔

jrsquared commented:
Some very cool features here! I can see how moving nodes will be very useful.

A few changes:
* Please revert file permission changes, all the files do not need execute privileges. 
* Ability to move nodes should be gated with default off, just like you did with physics. Using `setOptions()` and `showLabels` is a good example.
* You replaced the default layout worker with the ring layout. We need to keep the current layout worker for the `region` renderer.  Maybe layouts could be configurable the same way the renderer is configurable in the data structure, or a layout is tied to a renderer and the current default layout worker is tied to the `region` renderer. You could then create a new renderer type `ring` for your use case, and modify that in the traffic data structure. I'd be happy to help work on this change.
* If the orbitSize and node.size adjustments for GlobalTrafficGraph are important to you, please create a new graph type as well, the current sizes were chosen for a reason.

jrsquared commented on src/dns/dnsTrafficGraph.js:
```diff
@@ -124,11 +123,11 @@ class DNSTrafficGraph extends TrafficGraph {
       }
-      this.state.maxVolume = maxVolume * 1.5;
+      maxVolume *= 1.5;
+      this.state.maxVolume = maxVolume;
```
What's the difference?

jbrekelmans replied:
Some leftovers from development and playing around, no difference :)

jbrekelmans commented:
I like your suggestions.
Quite frankly they are essential given that I hardcoded a configuration that works for my graphs, but not necessarily for other graphs.
Here's my thoughts:
1. I agree with you when you say layouts should be configurable the same way renderer's are. We definitely shouldn't tie layouts and renderer's together from an ideal-design perspective, but they are tied together because only specific layout and renderer configurations work together. This is why I changed the orbit size and node size in the global renderer and probably is also the reason why the global renderer is tied together with it's layout. But tying together renderer's and layouts at the design level will only limit the possible configurations and there should be an option to specify the layout independently from the renderer. 
2. My changes to orbitSize and nodeSize are a form of hardcoding that will not work for arbitrary graphs. Anyway, they were chosen as such because in my graphs there are 6 regions instead of 3. Perhaps we can reconsider the calculation (well, introduce a calculation instead of using constants) that works for an arbitrary amount of regions?

jrsquared commented:
1. I agree specifying layout and renderers independently would be the best approach. I'll think more about this and get back to you. It will definitely require some refactoring of the current way layout works, maybe similar to how the renderers are architected...

2. In regards to orbitSize/nodeSize, I just pushed a commit to master (386ee4e) to do a better job figuring out the orbitSize. I lay out all the nodes, check to see what changes need to be made to orbitSize, and relayout if there is a new orbitSize.  Let me know if this satisfies your globalTrafficGraph needs...

jbrekelmans commented:
The new orbitSize calculation works for my graph.

jrsquared commented:
@jbrekelmans Take a look at 16ea932b5ffabdac0cd1f5918ed4e6c812249a83 in master.  I extracted the layouts from each renderer into a standard format (a class with a method with the signature `run(graph, dimensions, callback)`) and the ability to override the layout per renderer.

I think this would work for your use case as well, you would just move your layout code into a new RingLayout.js, follow the same pattern, and supply `layout: "ring"` in the JSON. Let me know what you think.

jbrekelmans commented:
The new code around layout a lot cleaner and my layout code is a lot simpler while doing the same thing. Nice! 

I'm not a big fan of passing classes around as variables, this makes sense for nodes and edges though in the base traffic graph. For layouts I would probably not pass the class but keep a number of instances/plugin-layouts in a central place. That said: I wouldn't spend time on this.

I have yet to polish out a few new features of mine, but I'll merge the master branches of vizceral-example and vizceral-react into mine and get back to you. 

In the meantime I've added support for historical data in my backend and have created a timeline/slider which allows me to query the last X time in windows of granularity Y: see screenshot below.

<img width="1179" alt="screen shot 2016-11-11 at 6 09 23 pm" src="https://cloud.githubusercontent.com/assets/4394224/20207364/cdaafcb0-a83a-11e6-8e25-e4ee1a0a4536.png">

jrsquared commented:
Looking at it again, I agree, passing the class around for layout is definitely not the best pattern. I also agree that not time worth spent re-architecting at this point until I get into that part of the code again. I'll keep it in mind for when I do work on that again though, thank you for pointing it out!

Very cool on the historical data... we are working on something related to historical data too.

jbrekelmans commented:
I've made a few small changes:
1. Support for small (zero or <100 maxVolume)
2. Added a check for null currentGraph in handleIntersectedObjectClick
3. Added/registered RingLayout

I don't have any more comments for this branch!

jasonwr commented:
@jbrekelmans There is already support for small (zero or <100 maxVolume) Also it doesn't make sense to have a maxVolume of 0. That would imply zero nodes (no graph).

jbrekelmans commented:
Hey Jason,
You are right: there is already support for both (whatever that means), but allow me to clarify: 
1. When dynamically changing maxVolume to zero any particles that have been released will no longer be animated but instead freeze. See commit 598f1ec line 440 for my change.
2. ConnectionView.updateVolume updates the ConnectionView.rateMap, which should be sorted by volume. If maxVolume is zero then linearRatio is Infinity. Do we want an entry in rateMap with Infinity released per tick??? Furthermore, if maxVolume is smaller than 100 the last entry will have a volume less than the second-to-last entry, which contradicts with the definition of this data structure. 

Did you look at the code?

jrsquared commented on src/base/trafficGraph.js:
```diff
@@ -427,10 +427,17 @@ class TrafficGraph extends EventEmitter {
+        let maxVolume = null;
+        if (hasOwnPropF.call(state, "maxVolume")) {
+          maxVolume = state.maxVolume;
+          if (typeof maxVolume === "number" && maxVolume % 1 === 0 && -1 < maxVolume && maxVolume < 1/0) {
```
Why does maxVolume need to be an integer (`maxVolume % 1 === 0`)?

jbrekelmans replied:
You're right, not needed

jrsquared commented on src/base/connectionView.js:
```diff
@@ -435,7 +437,7 @@ class ConnectionView extends BaseView {
 
     // We need the highest RPS connection to make this volume relative against
-    if (!this.object.volumeGreatest) { return; }
+    if (!this.object.volumeGreatest && this.object.volumeGreatest !== 0) { return; }
```
What are you trying to check for with this additional condition?
The condition will never hit anyways. if `this.object.volumeGreatest` exists, it will fail before checking the second condition.

jrsquared commented:
@jbrekelmans Great changes, We have not had a case where max volume is ever zero, but makes sense, and great fix.

jrsquared commented:
Looks like the only outstanding change would be to make movable node functionality optional, default being off.  It would probably fit well in `vizceral.setOptions()`, which currently only has `showLabels`.

jbrekelmans commented:
@jrsquared
My pleasure, added option to enable repositioning of nodes by dragging (disabled by default)

jrsquared commented:
Looks great! Fixed the one merge conflict that I introduced today. Great work.
